routemaster/services/autodrop.rb

```diff
@@ -0,0 +1,39 @@
+require 'routemaster/services'
+require 'routemaster/models/database'
+require 'routemaster/models/subscription'
+require 'routemaster/mixins/log'
+require 'ostruct'
+
+module Routemaster
+  module Services
+    # Autodrop messages from queues until the database is empty enough.
+    #
+    # Naive algorithm for now: just drop the oldest messages from all queues
+    # until we're below the low-water mark.
+    class Autodrop
+      include Mixins::Log
+
+      BATCH_SIZE = 100
+
+      def initialize
+        @database = Models::Database.instance
+      end
+
+      def call
+        _log.info { 'auto-drop: starting' }
+        return false unless @database.too_full?
+        messages_removed = 0
+        queues = Models::Subscription.map(&:queue)
+
+        # loop through queues, removing messages
+        until @database.empty_enough?
```

Comment:
Each iteration is _O(n)_ in Redis calls where _n_ is the number of queues. "Peeking" into a given queue is _O(1)_ as it uses [LINDEX -1](http://redis.io/commands/LINDEX):

> O(N) where N is the number of elements to traverse to get to the element at index. This makes asking for the first or the last element of the list O(1).

I did also do a quick stress-test; this code was able to remove about 20k messages/second, which is probably good enough for now?